### PR TITLE
Made SAM3 to return loss in outputs

### LINKED ISSUE #42659: SAM3 Loss
### Feature request

Hello,

Can you please add SAM3 to return loss in outputs (if target is provided)?

### Motivation

This feature would allow fine-tuning the SAM3 model for custom datasets.

### Your contribution

Currently I am not in a position to help due to my limited time, but it would be great to see this feature.

### PR DESCRIPTION
**Pull request by ash01825**
[CPU CI](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=46623) [GPU run-slow](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=46623)

# What does this PR do?
Fixes #42659

I have added the loss function for sam3 so now it gives loss as output when input with labels. Created a new Hungarianmatcher same as BinaryHungarianMatcherV2(https://github.com/facebookresearch/sam3/blob/main/sam3/train/matcher.py) which is the one currently used by sam3. Added loss functions and outputs include now auxiliary output, loss and loss_dict. Loss weights can be configured as per needs. Have added necessary tests and tested for them.

- [x] I confirm that this is not a pure code agent PR.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@molbap  
@yonigozlan

### CODE REVIEW

**Comment by molbap** [comment]
thanks for the PR @ash01825 ! Do you mind running `make style && make fix-repo` in a local install to make CI initial checks happy? then I'll take a look

**Comment by ash01825** [comment]
Hey @molbap  I ran `make style && make fix-repo` but it's leading to changes which aren't related to the issue and like in files I haven't even touched.
I kept the change which was related.

**Comment by ash01825** [comment]
@molbap @yonigozlan 
Sorry it was my issue I had updated ruff version being used.
The errors being thrown was because sam3_lite_text builds itself on the sam3 so had to add the loss for that too. 
Now it computes loss for sam3_lite_text as well. Also added configuration attributes for both sam3 and sam3_lite_text as we have an option for to edit loss weights.

**Comment by ash01825** [comment]
Hey @molbap I've passed the tests. Please review it when possible. Thanks!

**Comment by ash01825** [comment]
gentle ping... @molbap @guarin

File: src/transformers/loss/loss_sam3.py
Comment (molbap) [inline src/transformers/loss/loss_sam3.py:1]: Currently this bypasses the `LOSS_MAPPING` mechanism set in `loss_utils.py`

File: src/transformers/models/sam3/modeling_sam3.py [hunk 1]
```diff
@@ -2192,6 +2204,8 @@ def __init__(self, config: Sam3Config):
         # Dot product scoring to compute classification scores
         self.dot_product_scoring = Sam3DotProductScoring(config)
 
+        self._loss_function = Sam3Loss(config)
```
Comment (molbap) [inline src/transformers/models/sam3/modeling_sam3.py:2207]: this doesn't need to be set if `LOSS_MAPPING` is populated correctly (see other comment)

File: src/transformers/loss/loss_sam3.py [hunk 2]
```diff
@@ -0,0 +1,414 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+
+from ..image_transforms import corners_to_center_format
```
Comment (molbap) [inline src/transformers/loss/loss_sam3.py:19]: unguarded here, guarded later

File: src/transformers/loss/loss_sam3.py [hunk 3]
```diff
@@ -0,0 +1,414 @@
...
+    Args:
+        class_cost:
+            The relative weight of the classification error in the matching cost.
+        bbox_cost:
+            The relative weight of the L1 error of the bounding box coordinates in the matching cost.
+        giou_cost:
+            The relative weight of the giou loss of the bounding box in the matching cost.
+    """
+
+    def __init__(self, cost_class: float = 2.0, cost_bbox: float = 5.0, cost_giou: float = 2.0):
```
Comment (molbap) [inline src/transformers/loss/loss_sam3.py:59]: args are not matching docstring (also docstring names are more conventional)

File: src/transformers/loss/loss_sam3.py [hunk 4]
```diff
@@ -0,0 +1,414 @@
...
+class Sam3Loss(nn.Module):
+    """
+    This class computes the losses for Sam3Model. The process happens in two steps:
+        1) we compute Hungarian assignment between ground truth boxes and the outputs of the model
+        2) we supervise each pair of matched ground-truth / prediction (supervise class, boxes, and masks).
+    """
+
+    def __init__(self, config):
+        super().__init__()
+        self.matcher = Sam3HungarianMatcher()
```
Comment (molbap) [inline src/transformers/loss/loss_sam3.py:169]: config should be passed through here for loss to be configurable

File: src/transformers/loss/loss_sam3.py (hunk 3 above)
Comment (molbap) [inline src/transformers/loss/loss_sam3.py:59]: (also should be config as sole arg)

File: src/transformers/loss/loss_sam3.py [hunk 5]
```diff
@@ -0,0 +1,414 @@
...
+            n = sizes[b]
+            if n > 0:
+                cost_giou[b, :, :n] = -generalized_box_iou(
+                    center_to_corners_format(out_bbox[b]),
+                    center_to_corners_format(tgt_bbox[b, :n]),
+                )
+
+        # Focal binary class cost
+        alpha = 0.25
+        gamma = 2.0
```
Comment (molbap) [inline src/transformers/loss/loss_sam3.py:116]: this also should not be hardcoded but in config (likely ce_gamma, etc?)

File: tests/models/sam3/test_modeling_sam3.py [hunk 6]
```diff
@@ -276,6 +276,20 @@ def prepare_config_and_inputs(self):
...
 
+    def prepare_labels(self):
+        labels = []
+        for _ in range(self.batch_size):
+            num_boxes = 3
+            boxes = torch.rand(num_boxes, 4, device=torch_device)
+            boxes[:, 2:] = boxes[:, :2] + 0.2
+            boxes = torch.clamp(boxes, 0, 1)
+            masks = (
+                torch.rand(num_boxes, self.image_size // 4, self.image_size // 4, device=torch_device) > 0.5
```
Comment (molbap) [inline tests/models/sam3/test_modeling_sam3.py:287]: One small comment here would help!

File: src/transformers/loss/loss_sam3.py [hunk 7]
```diff
@@ -0,0 +1,414 @@
...
+            boxes = label["boxes"]  # xyxy
+            n = len(boxes)
+            num_boxes.append(n)
+            if n == 0:
+                continue
+            boxes_xyxy_list.append(boxes)
+            masks = label.get("masks")
+            if masks is not None:
+                has_masks = True
+                masks_list.append(masks)
```
Comment (molbap) [inline src/transformers/loss/loss_sam3.py:216]: then, can't we have issues with batches of images that are a mix of with and without masks?

**Review by molbap — COMMENTED** [review]
Hi @ash01825 , left a few comments!
Ideally, a test comparing a small `loss_dict` from sam3 original code would be perfect (we are currently just checking non-empty grads, etc). 
Also currently the loss creates an ImportError in envs without scipy, but that's linked to the LOSS_MAPPING not taken into account. LMK when you need another review :)

File: src/transformers/loss/loss_sam3.py (hunk 7 above)
Comment (ash01825) [reply src/transformers/loss/loss_sam3.py:216]: Hey I'll get on others this one like the official repo doesn't actually implement for mixed batches. So do we like add a throw error or track?
